test/e2e/framework/daemonset_util.go

```diff
@@ -0,0 +1,197 @@
+	return t.kc.AppsV1alpha1().DaemonSets(t.ns).Create(ds)
+}
+
+func (t *DaemonSetTester) GetDaemonSet(name string) (*appsv1alpha1.DaemonSet, error) {
+	return t.kc.AppsV1alpha1().DaemonSets(t.ns).Get(name, metav1.GetOptions{})
+}
+
+func (t *DaemonSetTester) UpdateDaemonSet(name string, fn func(ds *appsv1alpha1.DaemonSet)) error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		ds, err := t.GetDaemonSet(name)
+		if err != nil {
+			return err
+		}
+
+		fn(ds)
+		_, err = t.kc.AppsV1alpha1().DaemonSets(t.ns).Update(ds)
+		return err
+	})
+}
+
+func (t *DaemonSetTester) DeleteDaemonSet(namespace, name string) {
+	err := t.kc.AppsV1alpha1().DaemonSets(t.ns).Delete(name, &metav1.DeleteOptions{})
+	if err != nil {
+		Logf("delete daemonset(%s.%s) failed: %s", namespace, name, err.Error())
+		return
+	}
+}
+
+func (t *DaemonSetTester) WaitForDaemonSetDeleted(namespace, name string) {
+	pollErr := wait.PollImmediate(time.Second, time.Minute,
+		func() (bool, error) {
+			_, err := t.kc.AppsV1alpha1().DaemonSets(namespace).Get(name, metav1.GetOptions{})
+			if err != nil {
+				if errors.IsNotFound(err) {
+					return true, nil
+				}
+				return false, err
+			}
+			return false, nil
+		})
+	if pollErr != nil {
+		Failf("Failed waiting for daemonset to enter Deleted: %v", pollErr)
+	}
+}
+
+func (t *DaemonSetTester) CheckRunningOnAllNodes(ds *appsv1alpha1.DaemonSet) func() (bool, error) {
+	return func() (bool, error) {
+		nodeNames := t.SchedulableNodes(ds)
+		return t.CheckDaemonPodOnNodes(ds, nodeNames)()
+	}
+}
+
+func (t *DaemonSetTester) SchedulableNodes(ds *appsv1alpha1.DaemonSet) []string {
+	nodeList, err := t.c.CoreV1().Nodes().List(metav1.ListOptions{})
+	ExpectNoError(err)
+	nodeNames := make([]string, 0)
+	for _, node := range nodeList.Items {
+		if !t.CanScheduleOnNode(node, ds) {
+			Logf("DaemonSet pods can't tolerate node %s with taints %+v, skip checking this node", node.Name, node.Spec.Taints)
+			continue
+		}
+		nodeNames = append(nodeNames, node.Name)
+	}
+	return nodeNames
+}
+
+func (t *DaemonSetTester) CheckDaemonPodOnNodes(ds *appsv1alpha1.DaemonSet, nodeNames []string) func() (bool, error) {
+	return func() (bool, error) {
+		podList, err := t.c.CoreV1().Pods(t.ns).List(metav1.ListOptions{})
+		if err != nil {
+			Logf("could not get the pod list: %v", err)
+			return false, nil
+		}
+		pods := podList.Items
+
+		nodesToPodCount := make(map[string]int)
+		for _, pod := range pods {
+			if !metav1.IsControlledBy(&pod, ds) {
+				continue
+			}
+			if pod.DeletionTimestamp != nil {
+				continue
+			}
+			if podutil.IsPodAvailable(&pod, ds.Spec.MinReadySeconds, metav1.Now()) {
+				nodesToPodCount[pod.Spec.NodeName] += 1
+			}
+		}
+		Logf("Number of nodes with available pods: %d", len(nodesToPodCount))
+
+		// Ensure that exactly 1 pod is running on all nodes in nodeNames.
+		for _, nodeName := range nodeNames {
+			if nodesToPodCount[nodeName] != 1 {
```

Comment:
Should be `if nodesToPodCount[nodeName] > 1 ` ?

Reply:
Okay, I'll fix this.